.gitlab-ci.yml

```diff
@@ -1,48 +1,63 @@
 variables:
   GIT_SUBMODULE_STRATEGY: recursive
 
-cache:
-  key: ${CI_COMMIT_REF_SLUG}
-  paths:
-    - node_modules/
-    - dist/
-    - proto/
-    - docs/
-
 stages:
   - check
   - build
   - deploy
 
-before_script:
-  - node --version
-  - npm install
-
-prettier:
+lint:
+  image: registry.gitlab.com/matrixai/engineering/maintenance/gitlab-runner
   stage: check
-  image: node:12
   script:
-    - npm run lint
+    - nix-shell --packages nodejs --run "npm install"
+    - nix-shell --packages nodejs --run "npm run lint"
 
-jest:
-  stage: check
+test-npm:
   image: node:12
+  stage: check
   script:
+    - npm install
     - npm run test
 
-build_all:
+# # This requires QEMU to work on our image which currently does not work
+# test-nix:
+#   image: registry.gitlab.com/matrixai/engineering/maintenance/gitlab-runner
+#   stage: check
+#   script:
+#     - nix-build ./tests/nix/nat-testing.nix
+
+nix-dry:
+  image: registry.gitlab.com/matrixai/engineering/maintenance/gitlab-runner
+  stage: check
+  script:
+    - nix-build -v -v --dry-run ./release.nix --attr application
+    - nix-build -v -v --dry-run ./release.nix --attr docker
+
+nix:
+  image: registry.gitlab.com/matrixai/engineering/maintenance/gitlab-runner
   stage: build
-  image: node:12
   script:
-    - npm run build:all
+    - >
+      nix-build ./release.nix
+      --attr application
+      --attr docker
 
+build_npm:
+  image: registry.gitlab.com/matrixai/engineering/maintenance/gitlab-runner
+  stage: build
+  script:
+    - nix-shell --packages nodejs --run "npm install"
```

Comment:
I don't think we need to test `npm install`.

In fact I expect all builds to happen via `nix-build` instead. At least for application builds.

Reply:
this is not to test `npm install`, this is to test building via npm, this can pick up a lot of errors in the code actually, including import errors/type errors. But like you said, it probably will be picked up in `nix-build` anyway, so I guess we could remove it.

Reply:
oh and to be clear, the npm install is just to have the tools necessary to build available on path